src/poetry/core/packages/constraints/empty_constraint.py

```diff
@@ -1,38 +1,35 @@
-from typing import TYPE_CHECKING
-
 from poetry.core.packages.constraints.base_constraint import BaseConstraint
 
 
-if TYPE_CHECKING:
-    from poetry.core.packages.constraints import ConstraintTypes  # noqa
-
-
 class EmptyConstraint(BaseConstraint):
 
     pretty_string = None
 
-    def matches(self, _: "ConstraintTypes") -> bool:
+    def matches(self, _: "BaseConstraint") -> bool:
         return True
 
     def is_empty(self) -> bool:
         return True
 
-    def allows(self, other: "ConstraintTypes") -> bool:
+    def allows(self, other: "BaseConstraint") -> bool:
         return False
 
-    def allows_all(self, other: "ConstraintTypes") -> bool:
+    def allows_all(self, other: "BaseConstraint") -> bool:
         return True
 
-    def allows_any(self, other: "ConstraintTypes") -> bool:
+    def allows_any(self, other: "BaseConstraint") -> bool:
         return True
 
-    def intersect(self, other: "ConstraintTypes") -> "ConstraintTypes":
+    def intersect(self, other: "BaseConstraint") -> "BaseConstraint":
         return other
 
-    def difference(self, other: "ConstraintTypes") -> None:
-        return
+    def difference(self, other: "BaseConstraint") -> "BaseConstraint":
```

Comment:
I guess we can be more specific here by setting the return type to `EmptyConstraint`.

Reply:
We could, but that would be different from the signature on the BaseConstraint.  I prefer to keep the signatures in the various constraints consistent; but if you have strong opinions that it's useful to try and make them more precise in the variousl cases where this is possible... well, maybe another time?

Reply:
cf previous discussion https://github.com/python-poetry/poetry-core/pull/253/files#r757057577